### PR TITLE
feat: Add provider for Google Cloud Platform

### PR DESCRIPTION
**Pull request by fishkerez**
## Description

Add support for GCP provider. This is the code part, without the installation, which should come next.

## Type of Change

[ ] Bug Fix  
[ *] New Feature  
[ ] Breaking Change  
[ ] Refactor  
[ ] Documentation  
[ ] Other (please describe)  

## Checklist

- [* ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ *] Existing issues have been referenced (where applicable)
- [ *] I have verified this change is not present in other open pull requests
- [ *] Functionality is documented
- [ *] All code style checks pass
- [ *] New code contribution is covered by automated tests
- [ *] All new and existing tests pass

### CODE REVIEW

File: runtime_scan/pkg/provider/gcp/common.go
Comment (FrimIdan) [inline runtime_scan/pkg/provider/gcp/common.go:92]: maybe better to use `net/url` library?

https://go.dev/play/p/K_er3w-z5fr

https://stackoverflow.com/questions/59438561/how-to-obtain-the-last-path-segment-of-a-url

File: runtime_scan/pkg/provider/gcp/config.go
Comment (FrimIdan) [inline runtime_scan/pkg/provider/gcp/config.go:65]: where is `VMCLARITY_GCP_PROJECT_ID`, `VMCLARITY_GCP_SCANNER_VM_SIZE` and `VMCLARITY_GCP_SCANNER_SUBNETWORK` defined?

Also, `VMCLARITY_GCP_SCANNER_VM_SIZE` is not `VMCLARITY_GCP_SCANNER_ZONE`?

File: runtime_scan/pkg/provider/gcp/scannerVm.go
Comment (FrimIdan) [inline runtime_scan/pkg/provider/gcp/scannerVm.go:58]: ignore **not found** error OR ignore **not not found** errors? :)

File: runtime_scan/pkg/provider/gcp/scannerVm.go
Comment (FrimIdan) [inline runtime_scan/pkg/provider/gcp/scannerVm.go:93]: `10` is that a const? should be configurable?

File: runtime_scan/pkg/provider/gcp/snapshot.go
Comment (FrimIdan) [inline runtime_scan/pkg/provider/gcp/snapshot.go:64]: remove

File: runtime_scan/pkg/provider/gcp/client_test.go
Comment (chrisgacsal) [inline runtime_scan/pkg/provider/gcp/client_test.go:42]: I would use `gomega` as a testing library as it is more sophisticated and support comparing types where `reflect.DeepEqual` would fail. I'd move away from using `DeepEqual` for testing for sure. Not to mention that it also makes tests more readable.
Example for using `gomega`: https://github.com/openclarity/vmclarity/blob/main/runtime_scan/pkg/provider/cloudinit/cloudinit_test.go

**Review by chrisgacsal — CHANGES_REQUESTED** [review]

File: api/openapi.yaml
Comment (fishkerez) [reply api/openapi.yaml:1421]: us-central1-a/us-central1-b

File: runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go
Comment (fishkerez) [reply runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go:202]: @chrisgacsal The question is what if the error is NotFound? That is a legitimate error here and expected since we are looking to see if exists such scan, and if not we create it.

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:53]: Will change to validation failed

File: runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go
Comment (sambetts-cisco) [reply runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go:202]: There is no need to modify here, the code inside the provider should be handling any specific errors like not found and then either returning an error if its an error, a RetryableError if its to do requeued, or a FatalF if its fatal. GCP support should not touch *anything* outside of provider/gcp now

File: runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml
Comment (fishkerez) [reply runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml:20]: It is related to this PR since the constant /opt/vmclarity is not good for google container optimized image, since this part of the disk is read only

File: runtime_scan/pkg/provider/gcp/scannerVm.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/scannerVm.go:72]: Yes I believe so. Further more, we might need to dynamically look for an available machine type in the code base on the project/zone we are running at

File: runtime_scan/pkg/provider/gcp/client_test.go
Comment (sambetts-cisco) [reply runtime_scan/pkg/provider/gcp/client_test.go:42]: alternatively the standard we use across the rest of the project is cmp.Equals https://github.com/google/go-cmp it must nicer and doesn't pull in a ton of other dependencies

File: runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go
Comment (fishkerez) [reply runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go:202]: @sambetts-cisco I did not change anything as you can see. This is only a comment. Not sure I understand how NotFound error is being handled here.

File: runtime_scan/pkg/provider/gcp/scannerVm.go
Comment (sambetts-cisco) [reply runtime_scan/pkg/provider/gcp/scannerVm.go:74]: IMO this should be a configuration option as well, I'd like to avoid hard coding things like this so that we don't have to change the code if GCP renames something etc

File: runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go
Comment (sambetts-cisco) [reply runtime_scan/pkg/orchestrator/scanconfigwatcher/watcher.go:202]: This is a GET on endpoint api/scans, there should never be a "NotFound" unless the backend doesn't serve api/scans. If nothing matches the filter then you just get an response like `{ "Items": [] }`.

File: runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml
Comment (fishkerez) [reply runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml:20]: linux/ubuntu images on gcp does not have docker installed by default, The recommended image for gcp if you want to run containers is [Container-Optimized OS](https://cloud.google.com/container-optimized-os/docs)
This has docker and cloud init installed

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:198]: Since DiscoverScopes does not have a retry mechanism, I just return a simple error here

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:237]: The filtering in memory by tags is already implemented in this PR.

> And what do you mean tags filter is broken in the google api?

https://stackoverflow.com/questions/53417346/google-cloud-platform-how-to-filter-instances-list-by-tag

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:260]: There can be only one boot disk AFAIK

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:420]: No, here if tags are nil we return false (no exclude tags), this is because we don't want to exclude instance in the case that the user didn't provided any exclude tags
In hasIncludeTags, if tags are nil we return true, since no user tags mean it has include tags

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:452]: I am assuming that all vms will have at least one disk (boot disk)

File: runtime_scan/pkg/provider/gcp/scannerVm.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/scannerVm.go:58]: ignore not found. let it sync :)

File: runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml
Comment (fishkerez) [reply runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml:20]: Actually the ubuntu image that I checked (ubuntu-2004-lts) does not have cloud-init installed.
We can use startup scripts to install it, but not sure this is the approach we want

File: runtime_scan/pkg/provider/gcp/client_test.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client_test.go:25]: What do you expect to get in that case? key="tag3=" or key="tag3" ?

File: runtime_scan/pkg/provider/gcp/client.go
Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/client.go:260]: But it's a slice :)

File: runtime_scan/pkg/provider/gcp/client.go
Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/client.go:420]: I meant L409-L420
The convert and the for loop with the logic inside

File: runtime_scan/pkg/provider/gcp/client.go
Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/client.go:452]: I hope 👍

Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/client.go:452]: I hope 👍

File: runtime_scan/pkg/provider/gcp/config.go
Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/config.go:65]: > @FrimIdan Env vars defined here: https://github.com/openclarity/vmclarity/pull/433/files#diff-db8032f507629e5a7bb5e3459ed38e1ae146cb3e7dd2f56324465ab402407b06R39-R45

@chrisgacsal Problem that it's not a const so can't use the name here you then when changing the env var the print here will be wrong.
For example even now `VMCLARITY_GCP_SCANNER_VM_SIZE` is wrong I think, it should be `VMCLARITY_GCP_SCANNER_ZONE`, right?

File: runtime_scan/pkg/provider/gcp/client_test.go
Comment (chrisgacsal) [reply runtime_scan/pkg/provider/gcp/client_test.go:42]: `gomega` also relies on `go-cmp` in certain scenarios, but it provides much better DX in my opinion as it also includes matchers for other use-cases where neither `go-cmp` nor `DeepEqual` is appropriate.
See the docs: https://onsi.github.io/gomega/#provided-matchers

> it must nicer and doesn't pull in a ton of other dependencies

I'd not say that it does pull in a ton of deps: https://github.com/onsi/gomega/blob/master/go.mod
Actually it is quite slim compared to average go projects.

Anyways, we need to decide what are the tools we want to use for testing. Personally, I find both `go-cmp` and `reflect.DeppEqual` limited, so I do prefer using `gomega` for unit tests.

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:260]: vm.Disks is a slice that is true. But the boolean "Boot" can only be for one disk. The other disks should have this boolean false and are attached disks. Otherwise not sure how the the vm will know from which disk to boot

File: runtime_scan/pkg/provider/gcp/client.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client.go:420]: I can move it into a function

File: runtime_scan/pkg/provider/gcp/common.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/common.go:81]: tested as part of getZonesLastPart

File: runtime_scan/pkg/provider/gcp/config.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/config.go:65]: Will add consts for these

Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/config.go:65]: @FrimIdan switched to use consts

Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/config.go:65]: @fishkerez sure, they just all need to be aligned with the input to `BindEnv` and `SetEnvPrefix`
I mean maybe something like
```
const (
	DefaultEnvPrefix = "VMCLARITY_GCP"
	projectID = "project_id"
)

var projectIDEnvVar = strings.Join([]string{DefaultEnvPrefix, strings.ToUpper(projectID)}, "_")
.
.
.
	v.SetEnvPrefix(DefaultEnvPrefix)=

	_ = v.BindEnv(projectID)
```

Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/config.go:65]: Done

File: runtime_scan/pkg/provider/gcp/client_test.go
Comment (fishkerez) [reply runtime_scan/pkg/provider/gcp/client_test.go:42]: gomega looks very nice but for my use cases I think I can just stick with reflect.DeepEqual. 
Anyway this is what gomega.equal is doing

File: runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml
Comment (chrisgacsal) [reply runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml:20]: > It is related to this PR since the constant /opt/vmclarity is not good for google container optimized image, since this part of the disk is read only

I understand that, but this still could be a separate change merged prior this PR.

File: runtime_scan/pkg/provider/gcp/scannerVm.go
Comment (FrimIdan) [reply runtime_scan/pkg/provider/gcp/scannerVm.go:62]: my problem is that the comment is above `if !notFound {` so I'm a bit confused but it might just be me.

File: runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml
Comment (fishkerez) [reply runtime_scan/pkg/provider/cloudinit/cloud-init.tmpl.yaml]: Removed this for now until we decide if need this

**Review by sambetts-cisco — APPROVED** [review]
I've tested this E2E on GCP with the GCP installer and its working as expected. I'll approve however I contributed to the PR so I'd like to have at least one other approver before merging.

**Pull request merged**
